### PR TITLE
fix(notification): race detected send notification issue

### PR DESCRIPTION
**Pull request by msohailhussain**
### Summary
- Fixed an issue when notification callbacks were calling Add/Remove of same notification manager. 
- All summary is in this PR. https://github.com/optimizely/go-sdk/pull/219

### TestPlan
- Added unit tests for race detected scenarios.

### CODE REVIEW

File: pkg/notification/manager.go [hunk 1]
```diff
@@ -66,7 +66,17 @@ func (am *AtomicManager) Remove(id int) {
 
 // Send sends the notification to the registered handlers
 func (am *AtomicManager) Send(notification interface{}) {
-	for _, handler := range am.handlers {
+	handlers := am.copyHandlers()
+	for _, handler := range handlers {
```
Comment (pawels-optimizely) [inline pkg/notification/manager.go]: instead of copying everything at once, can we just do handler := handler inside the for loop ?

Comment (msohailhussain) [reply pkg/notification/manager.go]: tried didn't work.

File: pkg/notification/manager.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /****************************************************************************
- * Copyright 2019, Optimizely, Inc. and contributors                        *
+ * Copyright 2019-2020, Optimizely, Inc. and contributors                        *
```
Comment (mikeng13) [inline pkg/notification/manager.go]: Haha it's 2020! Still hasn't sunk in yet

File: pkg/notification/manager.go [hunk 3]
```diff
@@ -66,7 +74,20 @@ func (am *AtomicManager) Remove(id int) {
 
 // Send sends the notification to the registered handlers
 func (am *AtomicManager) Send(notification interface{}) {
-	for _, handler := range am.handlers {
+	handlers := am.copyHandlers()
+	for _, handler := range handlers {
+		// copying handler to avoid race condition
```
Comment (mikeng13) [inline pkg/notification/manager.go]: Move this comment to right before we copy handlers

File: pkg/notification/manager.go [hunk 4]
```diff
@@ -66,7 +74,20 @@ func (am *AtomicManager) Remove(id int) {
...
 func (am *AtomicManager) Send(notification interface{}) {
-	for _, handler := range am.handlers {
+	// copying handler to avoid race condition
+	handlers := am.copyHandlers()
+	for _, handler := range handlers {
 		handler(notification)
 	}
 }
+
+// Copy handlers and return it.
```
Comment (mikeng13) [inline pkg/notification/manager.go]: nit: Return a copy of the given handlers

File: pkg/notification/manager.go [hunk 5]
```diff
@@ -66,7 +74,20 @@ func (am *AtomicManager) Remove(id int) {
...
+func (am *AtomicManager) copyHandlers() map[uint32]func(interface{}) {
+	am.lock.RLock()
+	defer am.lock.RUnlock()
+	m := make(map[uint32]func(interface{}), len(am.handlers))
+	for k, v := range am.handlers {
+		m[k] = v
+	}
+	return m
+}
```
Comment (mikecdavis) [inline pkg/notification/manager.go]: We don't really need a map. We can simply copy to a list.

**Pull request merged**